File: CS/FilterNodeExample/MainForm.cs
using System;
using System.Linq;
using DevExpress.XtraEditors;
using DevExpress.XtraTreeList;
using System.Collections.Generic;
using DevExpress.XtraTreeList.Nodes;
using DevExpress.XtraTreeList.Columns;

namespace FilterNodeExample {
    public partial class MainForm :XtraForm {
        public MainForm() {
            InitializeComponent();
            TreeList.ExpandAll();
        }

        private void OnFilterNode(object sender, FilterNodeEventArgs e) {
            List<TreeListColumn> filteredColumns = e.Node.TreeList.Columns.Cast<TreeListColumn>(
                ).Where(c => c.FilterInfo.AutoFilterRowValue != null).ToList();
            if (filteredColumns.Count == 0) return;
            e.Handled = true;
            e.Node.Visible = filteredColumns.Any(c => IsNodeMatchFilter(e.Node, c));
        }

        static bool IsNodeMatchFilter(TreeListNode node, TreeListColumn column) {
            string filterValue = column.FilterInfo.AutoFilterRowValue.ToString();
            if (node.GetDisplayText(column).StartsWith(filterValue)) return true;
            foreach (TreeListNode n in node.Nodes)
                if (IsNodeMatchFilter(n, column)) return true;
            return false;
        }
    }
}

File: VB/FilterNodeExample/MainForm.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Linq
Imports DevExpress.XtraEditors
Imports DevExpress.XtraTreeList
Imports System.Collections.Generic
Imports DevExpress.XtraTreeList.Nodes
Imports DevExpress.XtraTreeList.Columns

Namespace FilterNodeExample
	Partial Public Class MainForm
		Inherits XtraForm
		Public Sub New()
			InitializeComponent()
			TreeList.ExpandAll()
		End Sub

		Private Sub OnFilterNode(ByVal sender As Object, ByVal e As FilterNodeEventArgs) Handles TreeList.FilterNode
			Dim filteredColumns As List(Of TreeListColumn) = e.Node.TreeList.Columns.Cast(Of TreeListColumn)().Where(Function(c) c.FilterInfo.AutoFilterRowValue IsNot Nothing).ToList()
			If filteredColumns.Count = 0 Then
				Return
			End If
			e.Handled = True
			e.Node.Visible = filteredColumns.Any(Function(c) IsNodeMatchFilter(e.Node, c))
		End Sub

		Private Shared Function IsNodeMatchFilter(ByVal node As TreeListNode, ByVal column As TreeListColumn) As Boolean
			Dim filterValue As String = column.FilterInfo.AutoFilterRowValue.ToString()
			If node.GetDisplayText(column).StartsWith(filterValue) Then
				Return True
			End If
			For Each n As TreeListNode In node.Nodes
				If IsNodeMatchFilter(n, column) Then
					Return True
				End If
			Next n
			Return False
		End Function
	End Class
End Namespace
